File: examples/05-durable-workflow/start_workflow.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 The Plinth Authors
"""Driver process for the durable-workflow demo.

This script:

1. Creates / fetches the ``durable-demo`` workspace.
2. Idempotently creates a ``research-pipeline`` workflow
   (manifest: ``search`` → ``fetch`` → ``extract`` → ``synth``).
3. Creates each step in ``initial_status="pending"`` so workers can
   lease them. (The legacy v0.2 default is ``running`` for in-process
   agents; the v0.5 durable executor wants pending steps.)
4. Polls the workflow until it transitions to ``completed`` /
   ``failed``, printing progress every few seconds.
5. Reads ``report.md`` from the workspace and prints it.

Run it AFTER starting at least one ``plinth-workflow-worker``:

    plinth-workflow-worker --handlers-module handlers --concurrency 2

You can kill the worker mid-flight (Ctrl+C); start another one — it
will pick up where the first left off.
"""

from __future__ import annotations

import argparse
import sys
import time

from plinth import Plinth

from shared import make_client_kwargs, services_available

WORKFLOW_NAME = "research-pipeline"
WORKFLOW_STEPS = ["search", "fetch", "extract", "synth"]


def _ensure_services() -> None:
    services = services_available()
    missing = [k for k, v in services.items() if not v]
    if missing:
        print(
            f"[start] services not reachable: {missing}. "
            "Start them with `make services` then retry.",
            file=sys.stderr,
        )
        sys.exit(2)


def _ensure_pending_steps(wf, *, topic: str) -> int:
    """Create any missing pending steps + return how many were started.

    Idempotent: re-running picks up where the previous attempt left off.
    Steps that already have a ``running`` / ``completed`` attempt are
    skipped.
    """

    wf.refresh()
    completed = {s.name for s in wf.steps if s.status == "completed"}
    in_flight = {s.name for s in wf.steps if s.status in {"running", "pending"}}
    started = 0
    for name in WORKFLOW_STEPS:
        if name in completed or name in in_flight:
            continue
        wf.start_step(
            name,
            input={"topic": topic, "k": 5},
            initial_status="pending",
        )
        started += 1
    return started


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(prog="start_workflow")
    parser.add_argument(
        "--topic",
        default="renewable energy",
        help="Research topic to drive the workflow.",
    )
    parser.add_argument(
        "--workspace-name",
        default="durable-demo",
        help="Name of the workspace (created if missing).",
    )
    parser.add_argument(
        "--timeout",
        type=int,
        default=120,
        help="Max seconds to wait for completion.",
    )
    parser.add_argument(
        "--poll-interval",
        type=float,
        default=2.0,
    )
    args = parser.parse_args(argv)

    _ensure_services()

    client = Plinth(**make_client_kwargs())
    ws = client.workspace(args.workspace_name)
    print(f"[start] workspace: {ws.id} ({ws.name})")

    wf = ws.workflows.get_or_create(WORKFLOW_NAME, steps=WORKFLOW_STEPS)
    print(f"[start] workflow: {wf.id} (status={wf.status})")

    started = _ensure_pending_steps(wf, topic=args.topic)
    if started:
        print(f"[start] queued {started} pending steps for the worker pool")
    else:
        print("[start] all steps already in flight or done")

    deadline = time.time() + args.timeout
    last_status: str | None = None
    while True:
        wf.refresh()
        status = wf.status
        if status != last_status:
            done = sum(1 for s in wf.steps if s.status == "completed")
            running = sum(1 for s in wf.steps if s.status == "running")
            pending = sum(1 for s in wf.steps if s.status == "pending")
            print(
                f"[start] status={status} "
                f"completed={done} running={running} pending={pending}"
            )
            last_status = status

        if status in {"completed", "failed", "cancelled"}:
            break
        if time.time() > deadline:
            print(f"[start] TIMEOUT after {args.timeout}s; current status={status}")
            return 1

        time.sleep(args.poll_interval)

    print(f"[start] workflow {wf.id} {status}")
    if status == "completed":
        try:
            report = ws.files.read("report.md", as_text=True)
        except Exception as exc:  # noqa: BLE001
            print(f"[start] could not read report.md: {exc}")
            return 0
        print("---- report.md ----")
        print(report)
    return 0 if status == "completed" else 1


if __name__ == "__main__":
    sys.exit(main())

File: examples/05-durable-workflow/start-workflow.ts
/**
 * SPDX-License-Identifier: Apache-2.0
 * Copyright 2026 The Plinth Authors
 *
 * Driver for the durable-workflow demo — TypeScript counterpart of
 * `start_workflow.py`.
 *
 * 1. Creates / fetches the `durable-demo` workspace.
 * 2. Idempotently creates a `research-pipeline` workflow with the
 *    manifest `search → fetch → extract → synth`.
 * 3. Creates each step in `pending` status so workers can lease them.
 * 4. Polls until the workflow transitions to `completed` / `failed`.
 * 5. Reads `report.md` from the workspace and prints it.
 *
 * Run AFTER starting at least one `plinth-workflow-worker`:
 *
 *     plinth-workflow-worker --handlers-module ./handlers.js --concurrency 2
 *
 * Then in another terminal:
 *
 *     node ./start-workflow.js --topic "renewable energy"
 *
 * You can kill the worker mid-flight; start another one — it will pick
 * up where the first left off.
 */

import { parseArgs } from "node:util";

import { Plinth, type WorkflowHandle } from "@plinth/sdk";

import { makeClientKwargs, servicesAvailable } from "./shared.js";

const WORKFLOW_NAME = "research-pipeline";
const WORKFLOW_STEPS = ["search", "fetch", "extract", "synth"];

async function ensureServices(): Promise<void> {
  const services = await servicesAvailable();
  const missing = Object.entries(services)
    .filter(([, ok]) => !ok)
    .map(([k]) => k);
  if (missing.length > 0) {
    process.stderr.write(
      `[start] services not reachable: ${missing.join(", ")}. ` +
        "Start them with `make services` then retry.\n",
    );
    process.exit(2);
  }
}

async function ensurePendingSteps(wf: WorkflowHandle, topic: string): Promise<number> {
  await wf.refresh();
  const completed = new Set(
    wf.steps.filter((s) => s.status === "completed").map((s) => s.name),
  );
  const inflight = new Set(
    wf.steps
      .filter((s) => s.status === "running" || s.status === "pending")
      .map((s) => s.name),
  );
  let started = 0;
  for (const name of WORKFLOW_STEPS) {
    if (completed.has(name) || inflight.has(name)) continue;
    // ``initialStatus: "pending"`` is the v0.5 opt-in: the step is
    // staged for a worker to lease rather than running in-process.
    await wf.startStep(name, {
      input: { topic, k: 5 },
      initialStatus: "pending",
    });
    started += 1;
  }
  return started;
}

async function main(): Promise<number> {
  const { values } = parseArgs({
    options: {
      topic: { type: "string", default: "renewable energy" },
      "workspace-name": { type: "string", default: "durable-demo" },
      timeout: { type: "string", default: "120" },
      "poll-interval": { type: "string", default: "2" },
    },
  });
  const topic = values.topic as string;
  const wsName = values["workspace-name"] as string;
  const timeout = Number.parseInt(values.timeout as string, 10);
  const pollInterval = Number.parseFloat(values["poll-interval"] as string) * 1000;

  await ensureServices();

  const cfg = makeClientKwargs();
  const client = new Plinth({
    workspaceUrl: cfg.workspaceUrl,
    gatewayUrl: cfg.gatewayUrl,
    apiKey: cfg.apiKey,
  });
  const ws = await client.workspace(wsName);
  process.stdout.write(`[start] workspace: ${ws.id} (${ws.name})\n`);

  const wf = await ws.workflows.getOrCreate(WORKFLOW_NAME, {
    steps: WORKFLOW_STEPS,
  });
  process.stdout.write(`[start] workflow: ${wf.id} (status=${wf.status})\n`);

  const started = await ensurePendingSteps(wf, topic);
  if (started > 0) {
    process.stdout.write(`[start] queued ${started} steps for the worker pool\n`);
  } else {
    process.stdout.write("[start] all steps already in flight or done\n");
  }

  const deadline = Date.now() + timeout * 1000;
  let lastStatus: string | null = null;
  // eslint-disable-next-line no-constant-condition
  while (true) {
    await wf.refresh();
    const status = wf.status;
    if (status !== lastStatus) {
      const done = wf.steps.filter((s) => s.status === "completed").length;
      const running = wf.steps.filter((s) => s.status === "running").length;
      const pending = wf.steps.filter((s) => s.status === "pending").length;
      process.stdout.write(
        `[start] status=${status} completed=${done} running=${running} pending=${pending}\n`,
      );
      lastStatus = status;
    }
    if (status === "completed" || status === "failed" || status === "cancelled") {
      process.stdout.write(`[start] workflow ${wf.id} ${status}\n`);
      if (status === "completed") {
        try {
          const report = await ws.files.readText("report.md");
          process.stdout.write("---- report.md ----\n");
          process.stdout.write(report);
          process.stdout.write("\n");
        } catch (err) {
          process.stdout.write(
            `[start] could not read report.md: ${(err as Error).message}\n`,
          );
        }
      }
      return status === "completed" ? 0 : 1;
    }
    if (Date.now() > deadline) {
      process.stdout.write(
        `[start] TIMEOUT after ${timeout}s; current status=${status}\n`,
      );
      return 1;
    }
    await new Promise<void>((r) => setTimeout(r, pollInterval));
  }
}

main()
  .then((code) => process.exit(code))
  .catch((err) => {
    process.stderr.write(`[start] error: ${(err as Error).message}\n`);
    process.exit(1);
  });
